File: java/ValidateBST.java
import java.util.*;

public class ValidateBST {
  // We need a helper function here so overload this method
  public static boolean validateBst(BST tree) {
    return validateBst(tree, Integer.MIN_VALUE, Integer.MAX_VALUE);
  }
	
	public static boolean validateBst(BST tree, int min, int max) {
		if (tree == null) return true;
		if (tree.value < min || tree.value >= max) return false;
		
		return validateBst(tree.left, min, tree.value) && validateBst(tree.right, tree.value, max);
		
	}

  static class BST {
    public int value;
    public BST left;
    public BST right;

    public BST(int value) {
      this.value = value;
    }
  }
}

File: dart/validate_bst.dart
import 'package:dart_numerics/dart_numerics.dart';

class BST {
  int value;
  BST left;
  BST right;
  
  BST(int value) {
    this.value = value;
  }
}

bool validateBst(BST tree) {
  return _validateBst(tree, int64MinValue, int64MaxValue);
}

/// [_validateBst] is a helper method that does all of the recursion for
/// [validateBst]
bool _validateBst(BST tree, int min, int max) {
  if (tree == null)  return null;
  if (tree.value < min || tree.value >= max) return false;
  return _validateBst(tree.left, min, tree.value) && _validateBst(tree.right, tree.value, max);
}